signalfx-java/src/main/java/com/signalfx/signalflow/WebSocketTransport.java

```diff
@@ -0,0 +1,592 @@
+                    if ((channel != null) && (!channel.isClosed())) {
+                        StreamMessage streamMessage = new StreamMessage("data", null,
+                                objectMapper.writeValueAsString(dataMap));
+                        channel.offer(streamMessage);
+                    } else {
+                        log.debug("ignoring message. channel not found {}",
+                                dataMessage.getChannelName());
+                    }
+                }
+            } catch (JsonProcessingException ex) {
+                log.error("failed to process messages", ex);
+            }
+        }
+
+        @Override
+        public void onMessage(String data) {
+
+            try {
+                @SuppressWarnings("unchecked")
+                Map<String, Object> dataMap = objectMapper.readValue(data, Map.class);
+
+                // Intercept KEEP_ALIVE messages
+                String event = (String) dataMap.get("event");
+                if ((event != null) && (event.equals("KEEP_ALIVE"))) {
+                    return;
+                }
+
+                event = (String) dataMap.get("type");
+                if (event != null) {
+                    // Authenticated messages inform us that our authentication has been accepted
+                    // and we can now consider the socket as "connected".
+                    if (event.equals("authenticated")) {
+                        log.info("WebSocket connection authenticated as {} (in {})",
+                                dataMap.get("userId"), dataMap.get("orgId"));
+                    } else {
+                        // All other messages should have a channel.
+                        String channelName = (String) dataMap.get("channel");
+                        if (channelName != null) {
+                            TransportChannel channel = channels.get(channelName);
+                            if ((channel != null) && (!channel.isClosed())) {
+                                StreamMessage message = new StreamMessage(event, null, data);
+                                channel.offer(message);
+                            } else {
+                                log.debug("ignoring message. channel not found {}", channelName);
+                            }
+                        }
+                    }
+                } else {
+                    log.debug("ignoring message. {}", dataMap);
+                }
+
+            } catch (IOException ex) {
+                log.error("failed to process messages", ex);
+            }
+        }
+
+        public void sendMessage(final Map<String, String> request) {
+
+            try {
+                String message = objectMapper.writeValueAsString(request);
+                this.connection.sendMessage(message);
+            } catch (IOException ex) {
+                throw new SignalFlowException("failed to send message", ex);
```

Comment:
the original caller here would always be the SignalFlowClient class (even on transport initial construction).  So the thrown runtime exception should prevent any transport or channel from existing.  And also be the correct response on other actions like stopping computation.  But I could force a complete transport close on any exception .. but I am not sure if that is the correct thing to do for all IOExceptions from the sendmessage.